# City Bike

### 1. Import data

Timespan: July 2019 - July 2020

In [3]:
#import libraries
import pandas as pd
import os

In [6]:
pwd

'/Users/younghash/Desktop/Hw'

In [10]:
df201906 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201906-citibike-tripdata.csv")
df201907 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201907-citibike-tripdata.csv")
df201908 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201908-citibike-tripdata.csv")
df201909 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201909-citibike-tripdata.csv")
df201910 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201910-citibike-tripdata.csv")
df201911 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201911-citibike-tripdata.csv")
df201912 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-201912-citibike-tripdata.csv")
df202001 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-202001-citibike-tripdata.csv")
df202002 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-202002-citibike-tripdata.csv")
df202003 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-202003-citibike-tripdata.csv")
df202004 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-202004-citibike-tripdata.csv")
df202005 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-202005-citibike-tripdata.csv")
df202006 = pd.read_csv("/Users/younghash/Desktop/Hw/Data/JC-202006-citibike-tripdata.csv")
df201906.head()

tripduration                 starttime                  stoptime  \
0           316  2019-06-01 00:05:55.7210  2019-06-01 00:11:11.9940   
1           280  2019-06-01 00:08:09.0920  2019-06-01 00:12:50.0730   
2          2162  2019-06-01 00:13:11.8800  2019-06-01 00:49:14.3640   
3          2184  2019-06-01 00:13:30.1110  2019-06-01 00:49:54.2250   
4           139  2019-06-01 00:13:31.8520  2019-06-01 00:15:51.7880   

   start station id start station name  start station latitude  \
0              3187          Warren St               40.721124   
1              3185          City Hall               40.717733   
2              3199       Newport Pkwy               40.728745   
3              3199       Newport Pkwy               40.728745   
4              3207        Oakland Ave               40.737604   

   start station longitude  end station id end station name  \
0               -74.038051            3199     Newport Pkwy   
1               -74.043845            3267     Morris Canal   
2               -74.032108            3199     Newport Pkwy   
3               -74.032108            3199     Newport Pkwy   
4               -74.052478            3207      Oakland Ave   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728745             -74.032108   29677  Subscriber   
1             40.712419             -74.038526   29571  Subscriber   
2             40.728745             -74.032108   26208    Customer   
3             40.728745             -74.032108   29487    Customer   
4             40.737604             -74.052478   29592  Subscriber   

   birth year  gender  
0        1993       1  
1        1984       1  
2        1969       0  
3        1969       0  
4        1994       1

In [11]:
df202006.head()

tripduration                 starttime                  stoptime  \
0           122  2020-06-01 00:02:21.7500  2020-06-01 00:04:24.3250   
1          1282  2020-06-01 00:12:27.1090  2020-06-01 00:33:50.0540   
2           186  2020-06-01 01:19:59.2090  2020-06-01 01:23:05.8930   
3           131  2020-06-01 01:26:42.2050  2020-06-01 01:28:53.8040   
4           266  2020-06-01 01:27:33.7090  2020-06-01 01:32:00.1890   

   start station id          start station name  start station latitude  \
0              3792  Columbus Dr at Exchange Pl               40.716870   
1              3198            Heights Elevator               40.748716   
2              3275              Columbus Drive               40.718355   
3              3267                Morris Canal               40.712419   
4              3639                  Harborside               40.719252   

   start station longitude  end station id  end station name  \
0               -74.032810            3639        Harborside   
1               -74.040443            3276  Marin Light Rail   
2               -74.038914            3681          Grand St   
3               -74.038526            3681          Grand St   
4               -74.034234            3186     Grove St PATH   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.719252             -74.034234   42489  Subscriber   
1             40.714584             -74.042817   42183  Subscriber   
2             40.715178             -74.037683   42140  Subscriber   
3             40.715178             -74.037683   42120  Subscriber   
4             40.719586             -74.043117   42437  Subscriber   

   birth year  gender  
0        1987       1  
1        1988       1  
2        1987       1  
3        1987       1  
4        1972       2

In [12]:
# Concatenate JC dataframes
df = pd.concat([df201906, df201907, df201908, df201909, df201910, df201911, df201912, \
           df202001, df202002, df202003, df202004, df202005, df202006])
df.head()

tripduration                 starttime                  stoptime  \
0           316  2019-06-01 00:05:55.7210  2019-06-01 00:11:11.9940   
1           280  2019-06-01 00:08:09.0920  2019-06-01 00:12:50.0730   
2          2162  2019-06-01 00:13:11.8800  2019-06-01 00:49:14.3640   
3          2184  2019-06-01 00:13:30.1110  2019-06-01 00:49:54.2250   
4           139  2019-06-01 00:13:31.8520  2019-06-01 00:15:51.7880   

   start station id start station name  start station latitude  \
0              3187          Warren St               40.721124   
1              3185          City Hall               40.717733   
2              3199       Newport Pkwy               40.728745   
3              3199       Newport Pkwy               40.728745   
4              3207        Oakland Ave               40.737604   

   start station longitude  end station id end station name  \
0               -74.038051            3199     Newport Pkwy   
1               -74.043845            3267     Morris Canal   
2               -74.032108            3199     Newport Pkwy   
3               -74.032108            3199     Newport Pkwy   
4               -74.052478            3207      Oakland Ave   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728745             -74.032108   29677  Subscriber   
1             40.712419             -74.038526   29571  Subscriber   
2             40.728745             -74.032108   26208    Customer   
3             40.728745             -74.032108   29487    Customer   
4             40.737604             -74.052478   29592  Subscriber   

   birth year  gender  
0        1993       1  
1        1984       1  
2        1969       0  
3        1969       0  
4        1994       1

### 2. Data exploration

In [13]:
df.describe()

tripduration  start station id  start station latitude  \
count  4.118760e+05     411876.000000           411876.000000   
mean   1.030295e+03       3304.189812               40.722542   
std    1.240443e+04        182.309139                0.007113   
min    6.100000e+01       3184.000000               40.709651   
25%    2.520000e+02       3195.000000               40.717733   
50%    3.990000e+02       3211.000000               40.721525   
75%    7.722500e+02       3276.000000               40.727224   
max    2.566420e+06       3792.000000               40.748716   

       start station longitude  end station id  end station latitude  \
count            411876.000000   411876.000000         411876.000000   
mean                -74.046437     3303.911626             40.722205   
std                   0.011216      191.324874              0.007015   
min                 -74.083639       83.000000             40.674181   
25%                 -74.050656     3193.000000             40.717733   
50%                 -74.043845     3209.000000             40.721124   
75%                 -74.038051     3276.000000             40.727224   
max                 -74.032108     3992.000000             40.842842   

       end station longitude         bikeid     birth year         gender  
count          411876.000000  411876.000000  411876.000000  411876.000000  
mean              -74.045956   32072.742860    1981.352465       1.149895  
std                 0.011173    6291.298732      10.505634       0.558393  
min               -74.083639   14531.000000    1887.000000       0.000000  
25%               -74.050389   29203.000000    1974.000000       1.000000  
50%               -74.043117   29504.000000    1984.000000       1.000000  
75%               -74.038051   41498.000000    1989.000000       1.000000  
max               -73.942125   44543.000000    2004.000000       2.000000

411876 trips have been recorded total during the chosen period.

In [15]:
pctg = round((len(df201906) / len(df202006) - 1)*100, 1)
print(f"From 2019-07 to 2020-07, total ridership has grown by {pctg}%.")

From 2019-07 to 2020-07, total ridership has grown by 6.8%.


In [16]:
sub201906 = df201906["usertype"].value_counts()["Subscriber"]
print("2019-06")
print(f"Proportion of subscribers: {sub201906 / df201906['usertype'].count()}")
cus201906 = df201906["usertype"].value_counts()["Customer"]
print(f"Proportion of customers: {cus201906 / df201906['usertype'].count()}")

sub202006 = df202006["usertype"].value_counts()["Subscriber"]
print("---\n2020-06")
print(f"Proportion of subscribers: {sub202006 / df202006['usertype'].count()}")
cus202006 = df202006["usertype"].value_counts()["Customer"]
print(f"Proportion of customers: {cus202006 / df202006['usertype'].count()}")

2019-06
Proportion of subscribers: 0.8536900836926198
Proportion of customers: 0.14630991630738016
---
2020-06
Proportion of subscribers: 0.5369572871807372
Proportion of customers: 0.46304271281926274


In [17]:
summer = pd.concat([df201906, df201907, df201908])
summer["hour"] = pd.to_datetime(summer.starttime)
summer["hour"] = summer["hour"].dt.hour
summer.groupby(["hour"])["bikeid"].count()

hour
0      1035
1       556
2       365
3       241
4       263
5      1210
6      3399
7      8374
8     13549
9      7793
10     5356
11     5239
12     6002
13     5848
14     5479
15     5734
16     7268
17    12254
18    13924
19    10370
20     7413
21     4863
22     3391
23     1961
Name: bikeid, dtype: int64

Peak hours are 8:00, 17:00, and 18:00 during summer months June, July, August.

In [18]:
winter = pd.concat([df201912, df202001, df202002])
winter["hour"] = pd.to_datetime(winter.starttime)
winter["hour"] = winter["hour"].dt.hour
winter.groupby(["hour"])["bikeid"].count()

hour
0      467
1      286
2      141
3       86
4      184
5      726
6     2127
7     4927
8     7972
9     4438
10    2999
11    3079
12    3243
13    3124
14    3021
15    3312
16    4129
17    6736
18    6753
19    4278
20    2707
21    1938
22    1260
23     777
Name: bikeid, dtype: int64

Peak hours are 8:00, 17:00, and 18:00 during winter months December, January, Febuary.

#### Today, top 10 stations in the city for starting a journey

In [27]:
# Query today's date from the dataframe. Closest date available since last upload by Citi Bike is June 30, 2020.
d='2020-06-30' 
today_date=pd.to_datetime(d).date()

df202006["date"] = pd.to_datetime(df202006["starttime"])
df202006["date"] = df202006["date"].dt.date
df_queried_date = df202006.loc[df202006["date"] == today_date]

In [29]:
# Find top stations
df_top_stations = df_queried_date.groupby(["start station id"])\
    ["start station id"].count().sort_values(ascending=False)[0:10].to_frame()
df_top_stations = df_top_stations.rename(columns={"start station id":"number of start stations"})
df_top_stations["start station id"] = df_top_stations.index
df_top_stations.reset_index(drop=True, inplace=True)
# df_top_stations

In [30]:
stations = df202006[["start station id", "start station name"]].drop_duplicates()
df_top_stations.merge(stations, how="inner", on="start station id")

number of start stations  start station id          start station name
0                        74              3199                Newport Pkwy
1                        62              3792  Columbus Dr at Exchange Pl
2                        58              3186               Grove St PATH
3                        49              3203               Hamilton Park
4                        45              3639                  Harborside
5                        39              3195                     Sip Ave
6                        39              3205           JC Medical Center
7                        38              3187                   Warren St
8                        36              3276            Marin Light Rail
9                        35              3193                Lincoln Park

#### Today, top 10 stations in the city for ending a journey

In [31]:
# Find top stations
df_top_stations = df_queried_date.groupby(["end station id"])\
    ["end station id"].count().sort_values(ascending=False)[0:10].to_frame()
df_top_stations = df_top_stations.rename(columns={"end station id":"number of end stations"})
df_top_stations["end station id"] = df_top_stations.index
df_top_stations.reset_index(drop=True, inplace=True)
# df_top_stations

In [24]:
stations = df202006[["end station id", "end station name"]].drop_duplicates()
df_top_stations.merge(stations, how="inner", on="end station id")

number of end stations  end station id            end station name
0                      77            3199                Newport Pkwy
1                      69            3186               Grove St PATH
2                      66            3792  Columbus Dr at Exchange Pl
3                      52            3203               Hamilton Park
4                      46            3639                  Harborside
5                      43            3205           JC Medical Center
6                      43            3195                     Sip Ave
7                      37            3187                   Warren St
8                      36            3276            Marin Light Rail
9                      35            3184                 Paulus Hook

#### Today, bottom 10 stations in the city for starting a journey

In [34]:
# Find bottom stations
df_bottom_stations = df_queried_date.groupby(["start station id"])\
    ["start station id"].count().sort_values(ascending=True)[0:10].to_frame()
df_bottom_stations = df_bottom_stations.rename(columns={"start station id":"number of start stations"})
df_bottom_stations["start station id"] = df_bottom_stations.index
df_bottom_stations.reset_index(drop=True, inplace=True)
# df_bottom_stations

In [35]:
stations = df202006[["start station id", "start station name"]].drop_duplicates()
df_bottom_stations.merge(stations, how="inner", on="start station id")

number of start stations  start station id   start station name
0                         2              3212      Christ Hospital
1                         3              3694       Jackson Square
2                         4              3220    5 Corners Library
3                         7              3201               Dey St
4                         8              3281  Leonard Gordon Park
5                         8              3207          Oakland Ave
6                         8              3640       Journal Square
7                         8              3198     Heights Elevator
8                         9              3679           Bergen Ave
9                        10              3677         Glenwood Ave

#### Today, bottom 10 stations in the city for ending a journey

In [37]:
# Find bottom stations
df_bottom_stations = df_queried_date.groupby(["end station id"])\
    ["end station id"].count().sort_values(ascending=True)[0:10].to_frame()
df_bottom_stations = df_bottom_stations.rename(columns={"end station id":"number of end stations"})
df_bottom_stations["end station id"] = df_bottom_stations.index
df_bottom_stations.reset_index(drop=True, inplace=True)
# df_bottom_stations

In [38]:
stations = df202006[["end station id", "end station name"]].drop_duplicates()
df_bottom_stations.merge(stations, how="inner", on="end station id")

number of end stations  end station id               end station name
0                       1             518                E 39 St & 2 Ave
1                       1            3664  North Moore St & Greenwich St
2                       4            3201                         Dey St
3                       4            3220              5 Corners Library
4                       4            3212                Christ Hospital
5                       5            3207                    Oakland Ave
6                       5            3198               Heights Elevator
7                       5            3677                   Glenwood Ave
8                       7            3678                  Fairmount Ave
9                       8            3640                 Journal Square

### 3. Save combined data

In [39]:
df.to_csv("past_year.csv")